solr/core/src/test/org/apache/solr/cloud/overseer/ZkStateReaderTest.java

```diff
@@ -644,28 +644,32 @@ public void testDeletePrsCollection() throws Exception {
         () -> reader.getClusterState().getCollectionOrNull("c1") != null);
 
     String collectionPath = ZkStateReader.getCollectionPath("c1");
-    ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(new SolrNamedThreadFactory("zk-state-reader-test-scheduler"));
-    executor.schedule(() -> {
-      try {
-        reader.getZkClient().clean(collectionPath); //call this directly, very tricky to control execution order with writer.enqueueUpdate
-      } catch (InterruptedException e) {
-        throw new RuntimeException(e);
-      } catch (KeeperException e) {
-        throw new RuntimeException(e);
-      }
-    },100, TimeUnit.MILLISECONDS);
 
-    CommonTestInjection.setDelay(1000);
-    timeOut.waitFor("Timeout waiting for collection state to become null", () -> {
-          //this should not throw exception even if the PRS entry read is delayed artificially (by previous command)
-          //and deleted after the following fetchCollectionState call
-      try {
-        return reader.fetchCollectionState("c1", null, ZkStateReader.getCollectionPath("c1")) == null;
-      } catch (KeeperException e) {
-        throw new RuntimeException(e);
-      } catch (InterruptedException e) {
-        throw new RuntimeException(e);
-      }
-    });
+    try {
+      CommonTestInjection.setWaitForCondition(ZkStateReader.class.getName(), () -> {
+        try {
+          //this is invoked after ZkStateReader.fetchCollectionState has fetched the state.json but before PRS entries
+          reader.getZkClient().clean(collectionPath); //call delete state.json on ZK directly, very tricky to control execution order with writer.enqueueUpdate
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        } catch (KeeperException e) {
+          throw new RuntimeException(e);
+        }
+      });
+
+      timeOut.waitFor("Timeout waiting for collection state to become null", () -> {
+        //this should not throw exception even if the PRS entry read is delayed artificially (by previous command)
+        //and deleted after the following fetchCollectionState call
```

Comment:
heck, with this injection framework you could actually verify that we exercise the case we're interested in, I think?:
```java
// in test
final boolean[] exercised = new boolean[1];
CommonTestInjection.setInject(ZkStateReader.class.getName()+"/exercised", () -> {
  exercised[0] = true;
});
...
// in replicaStateProvider:
...
} catch (PerReplicaStates.PrsZkNodeNotFoundException e) {
  assert CommonTestInjection.inject(ZkStateReader.class.getName()+"/exercised");
  throw e; //Do not wrap this into a generic SolrException
} catch (Exception e) {
...
// back in test, at end
assertTrue(exercised[0]);
```